streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -364,6 +370,73 @@ public void setUncaughtExceptionHandler(final Thread.UncaughtExceptionHandler eh
         }
     }
 
+    /**
+     * Set the handler invoked when a {@link StreamsConfig#NUM_STREAM_THREADS_CONFIG internal thread}
+     * throws an unexpected exception.
+     * These might be exceptions indicating rare bugs in Kafka Streams, or they
+     * might be exceptions thrown by your code, for example a NullPointerException thrown from your processor
+     * logic.
+     * <p>
+     * Note, this handler must be threadsafe, since it will be shared among all threads, and invoked from any
+     * thread that encounters such an exception.
+     *
+     * @param streamsUncaughtExceptionHandler the uncaught exception handler of type {@link StreamsUncaughtExceptionHandler} for all internal threads; {@code null} deletes the current handler
+     * @throws IllegalStateException if this {@code KafkaStreams} instance is not in state {@link State#CREATED CREATED}.
+     */
+    public void setUncaughtExceptionHandler(final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler) {
+        final StreamsUncaughtExceptionHandler handler = throwable -> handleStreamsUncaughtException(throwable, streamsUncaughtExceptionHandler);
+        synchronized (stateLock) {
+            if (state == State.CREATED) {
+                for (final StreamThread thread : threads) {
```

Comment:
What's our plan for the global thread?  I didn't think of this during the KIP discussion, and sorry if it was brought up there and I just forgot about it. But it seems like we should still give users a non-deprecated way to set a handler for the global thread.

Reply:
If we want to use the same StreamsUncaughtExceptionHandler so users don't have to implement two different handlers, which sounds right to me, then maybe we should add a parameter to the `handle` method to indicate whether it's a global or stream thread. Or split into a separate `handleStreamThreadException` and `handleGlobalThreadException` methods, or something like that (and possibly rename `SHUTDOWN_STREAM_THREAD` to `SHUTDOWN_THREAD`)
WDYT?

Reply:
I think users may find at least some of the same functionality to be useful, in particular the "shutdown client" and "shutdown application" enums. I also feel like users may want to be able to restart the global thread.  I think we can consider that out of scope for now, but I'd prefer to avoid introducing a new method that just accepts the old kind of uncaught exception handler if we're just going to deprecate that too.

But I do agree that it's pretty different, and we also don't want to commit to adding this functionality for the global thread right away so we should make sure it's clear what is and isn't implemented. So what do you think about mirroring the new handler for the StreamThread, eg adding a "GlobalUncaughtExceptionHandler", and just only including the `SHUTDOWN_THREAD` enum in the initial version? That way we're set up to easily extend the global handling feature without necessarily needing to implement it right away